Stats with Opentelemetry (#141)
deeppavlov/dialog_flow_framework: docs/source/user_guides/superset_guide.rst

    docker-compose up otelcol clickhouse dashboard

**Collecting data**

Collecting data is done by means of instrumenting your conversational service before you run it.
DFF tutorials showcase all the necessary steps, needed to achieve that. We will run
`one of those files <../tutorials/tutorials.stats.1_extractor_functions.py>`_ 
in order to obtain sample data points to visualize.

.. code-block:: shell

    export DISABLE_INTERACTIVE_MODE=1 && python tutorials/stats/1_extractor_functions.py

Displaying the data
~~~~~~~~~~~~~~~~~~~

In order to display the Superset dashboard, you should update the default configuration with the credentials of your database.
The configuration can be optionally saved as a zip archive for inspection / debug.

You can set most of the configuration options using a YAML file.

.. code-block:: yaml
    :linenos:

    # config.yaml
    db:
        type: clickhousedb+connect
        name: test
        user: username
        host: clickhouse
        port: 8123
        table: otel_logs

The file can then be used to parametrize the configuration script.
Password values will be prompted once you run the following command:

.. code-block:: shell

>     dff.stats config.yaml -P -dP -U superset --outfile=config_artifact.zip

Reviewer: I guess this command should be also able to pick up passwords from env variables, don't you think?

Author: Since it's meant to be agnostic to database type, it's not clear which environment variable it should use for that. We know the variable name for Superset, but it would be strange to pick only one password from the environment

Reviewer: Fair enough; do we plan any other databases, not superet?

Author: Not at the moment, but I still see no reason to hardcode it. For instance, users can pass the environment variables to the cli interface directly if they need to.

Superset guide
---------------------

Description
~~~~~~~~~~~

| The Dialog Flow Stats module can be used to obtain and visualize usage statistics for your service.
| The module relies on several open source solutions that allow for data persistence and visualization

* `Clickhouse <https://clickhouse.com/>`_ serves as an OLAP storage for data.
* Batching and preprocessing data is based on `OpenTelemetry protocol <https://opentelemetry.io/docs/what-is-opentelemetry/>`_ and the `OpenTelemetry collector <https://opentelemetry.io/docs/collector/>`_.
* Interactive visualization is powered by `Apache Superset <https://superset.apache.org/>`_.

All the mentioned services are shipped as Docker containers, including a pre-built Superset image that ensures API compatibility.

Collection procedure
~~~~~~~~~~~~~~~~~~~~

**Installation**

.. code-block:: shell

    pip install dff[stats]

**Launching services**

.. code-block:: shell
    :linenos:

    # clone the original repository to access the docker-compose file
    git clone https://github.com/deeppavlov/dialog_flow_framework.git
    # launch the required services

>     docker-compose up otelcol clickhouse dashboard

Author: added

**Collecting data**

Collecting data is done by means of instrumenting your conversational service before you run it.
DFF tutorials showcase all the necessary steps, needed to achieve that. We will run
`one of those files <../tutorials/tutorials.stats.1_extractor_functions.py>`_ 
in order to obtain sample data points to visualize.

.. code-block:: shell

    export DISABLE_INTERACTIVE_MODE=1 && python tutorials/stats/1_extractor_functions.py

Displaying the data
~~~~~~~~~~~~~~~~~~~

In order to display the Superset dashboard, you should update the default configuration with the credentials of your database.
The configuration can be optionally saved as a zip archive for inspection / debug.

You can set most of the configuration options using a YAML file.

.. code-block:: yaml
    :linenos:

    # config.yaml
    db:
        type: clickhousedb+connect
        name: test
        user: username
        host: clickhouse
        port: 8123
        table: otel_logs

The file can then be used to parametrize the configuration script.

> Password values will be prompted once you run the following command:
> 
> .. code-block:: shell
> 
>     dff.stats tutorials/stats/config.yaml -P -dP -U superset --outfile=config_artifact.zip

Reviewer 2: I think we should just pass passwords via CLI:
```suggestion

.. code-block:: shell

    dff.stats tutorials/stats/config.yaml -P superset -dP pass -U superset --outfile=config_artifact.zip

.. warning::
    
    Here we passed passwords via CLI, which is not recommended. For enhanced security, call the command above omitting the passwords (`dff.stats -P -dP -U superset ...`) and you will be prompted to enter them interactively.
```
That way, users don't have to guess them / look them up.

>     dff.stats tutorials/stats/config.yaml -P -dP -U superset --outfile=config_artifact.zip

Author: I think, it would make more sense to demonstrate the yaml syntax than the file creation process; Especially given that we have a ready file in the 'tutorials' section.

>     dff.stats tutorials/stats/config.yaml -P -dP -U superset --outfile=config_artifact.zip

Reviewer 2: I think that file creation process also shows the syntax of the file, but OK.

I like the change from
```
.. code-block:: yaml
    :linenos:

    # config.yaml
```
to
```
.. code-block:: yaml
    :linenos:

    # tutorials/stats/config.yaml
```
but I think it needs more explanation.

Something like
```
You can set most of the configuration options using a YAML file.
You can find a configuration file that works with the default env variables
in the `tutorials/stats` directory:

.. code-block:: yaml
    :linenos:

    # tutorials/stats/config.yaml
```

Running the command will automatically import the dashboard as well as the data sources
into the running superset server. If you are using a version of Superset different from the one
shipped with DFF, make sure that your access rights are sufficient to edit the workspace.

Using Superset
~~~~~~~~~~~~~~

> | In order to view the imported dashboard, log into `Superset <http://localhost:8088/>`_ using your username and password.

Reviewer 2: ```suggestion
| In order to view the imported dashboard, log into `Superset <http://localhost:8088/>`_ using your username and password (which are both `superset` by default and can be configured via `.env_file`).
```

Superset guide
---------------------

Description
~~~~~~~~~~~

| The Dialog Flow Stats module can be used to obtain and visualize usage statistics for your service.
| Data aggregation relies on the `OpenTelemetry protocol <#>`_ and the `OpenTelemetry collector <#>`_ along with `Clickhouse <https://clickhouse.com/>`_ as an OLAP storage.
| Interactive visualization is powered by `Apache Superset <https://superset.apache.org/>`_.
| All the mentioned services are shipped as Docker containers, including a pre-built Superset image that ensures API compatibility.
| Authorization credentials can be automatically configured through environment variables.

.. code-block:: shell

    echo 'SUPERSET_USERNAME=...' >> .env
    echo 'SUPERSET_PASSWORD=...' >> .env

>     docker run --env-file='.env' --name=superset ghcr.io/deeppavlov/superset_df_dashboard:latest

Author: Removed the lines

Collection procedure
~~~~~~~~~~~~~~~~~~~~

**Installation**

.. code-block:: shell

    pip install dff[stats]

**Launching services**

.. code-block:: shell

    docker-compose up otelcol clickhouse dashboard

**Collecting data**

Collecting data is done by means of instrumenting your conversational service before you run it.
DFF tutorials showcase all the necessary steps, needed to achieve that. We will run

> `one of those files <https://deeppavlov.github.io/dialog_flow_framework/tutorials/tutorials.stats.1_extractor_functions.html/>`_ 

Reviewer 2: ```suggestion
`one of those files <../tutorials/tutorials.stats.1_extractor_functions.py>`_ 
```

in order to obtain sample data points to visualize.

.. code-block:: shell

    export DISABLE_INTERACTIVE_MODE=1 && python tutorials/stats/1_extractor_functions.py

Displaying the data
~~~~~~~~~~~~~~~~~~~

In order to display the Superset dashboard, you should update the default configuration with the credentials of your database.
The configuration can be optionally saved as a zip archive for inspection / debug.

You can set most of the configuration options using a YAML file.

.. code-block:: yaml
    :linenos:

    # config.yaml
    db:
        type: clickhousedb+connect
        name: test
        user: username
        host: clickhouse
        port: 8123
        table: otel_logs

The file can then be used to parametrize the configuration script.
Password values can be omitted and set interactively.

.. code-block:: shell
    :linenos:

    dff.stats config.yaml -P -dP \
    -U superset \
    --db.type=clickhousedb+connect \
    --db.user=username \
    --db.host=localhost \
    --db.port=8123 \
    --db.name=test \
    --db.table=otel_logs \

>     --outfile=config_artifact.zip

Author: Changed

Superset configuration
=======================

Description
-----------

| The Dialog Flow Stats module can be used to obtain and visualize usage statistics for your service.
| Data aggregation relies on the `OpenTelemetry protocol <#>`_ and the `OpenTelemetry collector <#>`_ along with `Clickhouse <https://clickhouse.com/>`_ as an OLAP storage.
| Interactive visualization is powered by `Apache Superset <https://superset.apache.org/>`_.
| All the mentioned services are shipped as Docker containers, including a pre-built Superset image that ensures API compatibility.
| Authorization credentials can be automatically configured through environment variables.

.. code-block:: shell

    echo 'SUPERSET_USERNAME=...' >> .env
    echo 'SUPERSET_PASSWORD=...' >> .env
    docker run --env-file='.env' ghcr.io/deeppavlov/superset_df_dashboard:latest

Collection procedure
--------------------

**Installation**

.. code-block:: shell

    pip install dff[stats,clickhouse]
    pip install dff[stats,postgresql]

**Launching services**

.. code-block:: shell

>     docker-compose up

Author: Done

**Setting up a pipeline**

It is essential that you use get_current_label at least once, so that the default Superset charts
can be successfully displayed.

.. code-block:: python
    :linenos:

    # import dependencies
    from dff.pipeline import Pipeline, Service, ACTOR
    from dff.stats import default_extractors
    # using default_extractors.current_label is required for dashboard integration 
    from dff.stats import OtelInstrumentor, set_logger_destination, set_tracer_destination
    from dff.stats import OTLPLogExporter, OTLPSpanExporter
    # initialize opentelemetry
    # insecure parameter ensures that SSL encryption is not forced
    dff_instrumentor = OtelInstrumentor.from_url("grpc://localhost:4317", insecure=True)
    dff_instrumentor.instrument()
    # Special extractor functions can be used as handlers to report statistics for any pipeline component.
    pipeline = Pipeline.from_dict(
        {
            "components": [
                Service(
                    handler=ACTOR,
                    after_handler=[
                        default_extractors.get_current_label, # using required extractor
                    ],
                ),

>             ]

Author: Done

:orphan:

Superset guide
---------------------

Description
~~~~~~~~~~~

| The Dialog Flow Stats module can be used to obtain and visualize usage statistics for your service.
| Data aggregation relies on the `OpenTelemetry protocol <#>`_ and the `OpenTelemetry collector <#>`_ along with `Clickhouse <https://clickhouse.com/>`_ as an OLAP storage.
| Interactive visualization is powered by `Apache Superset <https://superset.apache.org/>`_.
| All the mentioned services are shipped as Docker containers, including a pre-built Superset image that ensures API compatibility.
| Authorization credentials can be automatically configured through environment variables.

.. code-block:: shell

    echo 'SUPERSET_USERNAME=...' >> .env
    echo 'SUPERSET_PASSWORD=...' >> .env
    docker run --env-file='.env' ghcr.io/deeppavlov/superset_df_dashboard:latest

Collection procedure
~~~~~~~~~~~~~~~~~~~~

**Installation**

.. code-block:: shell

    pip install dff[stats,clickhouse]

>     pip install dff[stats,postgresql]

Reviewer 2: Also, installing `clickhouse` and `postgresql` seems unnecessary. `clickhouse` contains requirements for testing clickhouse, which user doesn't need, and `postgresql` contains requirements for context storage which is not used here.

>     pip install dff[stats,postgresql]

Author: Removed postgres, it's not supported anymore
